File: reference/gaze-corrector/behavior_fsm.py
"""Natural behavior state machine with 4 states and hysteresis."""

import time
from enum import Enum, auto

import config
from gaze_estimator import GazeInfo


class State(Enum):
    ENGAGED = auto()
    DISENGAGING = auto()
    DISENGAGED = auto()
    RE_ENGAGING = auto()


class BehaviorFSM:
    """Determines whether gaze correction should be active based on user behavior.

    Uses hysteresis thresholds and timed transitions to avoid flickering.
    """

    def __init__(self):
        self.state = State.ENGAGED
        self._transition_start: float = 0.0

    def _should_force_disengage(self, gaze: GazeInfo) -> bool:
        """Head pose override: always disengage for large head turns."""
        return (abs(gaze.yaw) > config.HEAD_YAW_THRESHOLD or
                abs(gaze.pitch) > config.HEAD_PITCH_THRESHOLD)

    def _gaze_is_away(self, gaze: GazeInfo) -> bool:
        return gaze.gaze_angle > config.DISENGAGE_THRESHOLD

    def _gaze_is_near(self, gaze: GazeInfo) -> bool:
        return gaze.gaze_angle < config.ENGAGE_THRESHOLD

    def update(self, gaze: GazeInfo) -> float:
        """Update FSM and return correction blend factor (0.0 = off, 1.0 = full).

        Args:
            gaze: Current frame's gaze estimation.

        Returns:
            Blend factor in [0.0, 1.0].
        """
        now = time.monotonic()
        force_away = self._should_force_disengage(gaze)

        if self.state == State.ENGAGED:
            if force_away or self._gaze_is_away(gaze):
                self.state = State.DISENGAGING
                self._transition_start = now
                return 1.0
            return 1.0

        elif self.state == State.DISENGAGING:
            elapsed = now - self._transition_start
            if not force_away and self._gaze_is_near(gaze):
                # Returned quickly, go back to engaged
                self.state = State.ENGAGED
                return 1.0
            if elapsed >= config.DISENGAGE_DURATION:
                self.state = State.DISENGAGED
                return 0.0
            # Linear fade out
            return max(0.0, 1.0 - elapsed / config.DISENGAGE_DURATION)

        elif self.state == State.DISENGAGED:
            if not force_away and self._gaze_is_near(gaze):
                self.state = State.RE_ENGAGING
                self._transition_start = now
                return 0.0
            return 0.0

        elif self.state == State.RE_ENGAGING:
            elapsed = now - self._transition_start
            if force_away or self._gaze_is_away(gaze):
                self.state = State.DISENGAGED
                return 0.0
            if elapsed >= config.RE_ENGAGE_DURATION:
                self.state = State.ENGAGED
                return 1.0
            # Linear fade in
            return min(1.0, elapsed / config.RE_ENGAGE_DURATION)

        return 0.0  # Fallback

File: apps/macos/EyesOn/Core/BehaviorFSM.swift
import Foundation

// Natural-behaviour state machine.
// Ported from `reference/gaze-corrector/behavior_fsm.py` (MIT).
//
// This is the single most important piece for making the product feel natural.
// If the user is genuinely looking at their notes, the correction must withdraw;
// when they come back, it must return smoothly. Forcing permanent eye contact is
// unsettling — worse than not correcting at all.
//
//      gaze near                gaze away
//    ┌──────────┐            ┌──────────────┐
//    │ ENGAGED  │───────────►│ DISENGAGING  │──── 0.4 s ────► ┌────────────┐
//    │ blend=1  │◄───────────│  (fading)    │                 │ DISENGAGED │
//    └──────────┘            └──────────────┘                 │  blend=0   │
//          ▲                                                  └─────┬──────┘
//          │                  ┌──────────────┐                      │
//          └──── 0.2 s ───────│ RE_ENGAGING  │◄──── gaze near ──────┘
//                             │  (fading in) │
//                             └──────────────┘

enum BehaviorState: String {
    case engaged     = "ENGAGED"
    case disengaging = "DISENGAGING"
    case disengaged  = "DISENGAGED"
    case reEngaging  = "RE_ENGAGING"
}

/// Decides *whether* and *how much* correction to apply, based on user behaviour.
///
/// Returns a single number: `blend ∈ [0, 1]`, which scales the correction magnitude.
/// The clock is injected so the transitions can be unit-tested deterministically.
struct BehaviorFSM {

    private(set) var state: BehaviorState = .engaged
    private var transitionStart: TimeInterval = 0

    /// Head turned or tilted far enough that correction is never appropriate.
    private func forcesDisengage(_ pose: HeadPose) -> Bool {
        abs(pose.yawDeg)   > CorrectionConfig.headYawThresholdDeg ||
        abs(pose.pitchDeg) > CorrectionConfig.headPitchThresholdDeg
    }

    private func gazeIsAway(_ angleDeg: Double) -> Bool {
        angleDeg > CorrectionConfig.disengageThresholdDeg
    }

    private func gazeIsNear(_ angleDeg: Double) -> Bool {
        angleDeg < CorrectionConfig.engageThresholdDeg
    }

    /// Advances the machine one frame and returns the blend factor.
    ///
    /// - Parameters:
    ///   - gazeAngleDeg: current gaze deviation from the camera, in degrees
    ///   - pose: current head orientation
    ///   - now: monotonic time (`CACurrentMediaTime()` in the app, injectable in tests)
    mutating func update(gazeAngleDeg: Double, pose: HeadPose, now: TimeInterval) -> Double {
        let forceAway = forcesDisengage(pose)

        switch state {
        case .engaged:
            if forceAway || gazeIsAway(gazeAngleDeg) {
                state = .disengaging
                transitionStart = now
            }
            return 1.0

        case .disengaging:
            let elapsed = now - transitionStart
            if !forceAway && gazeIsNear(gazeAngleDeg) {
                // Came back quickly — snap back to engaged rather than finishing the fade.
                state = .engaged
                return 1.0
            }
            if elapsed >= CorrectionConfig.disengageDuration {
                state = .disengaged
                return 0.0
            }
            return max(0.0, 1.0 - elapsed / CorrectionConfig.disengageDuration)

        case .disengaged:
            if !forceAway && gazeIsNear(gazeAngleDeg) {
                state = .reEngaging
                transitionStart = now
            }
            return 0.0

        case .reEngaging:
            let elapsed = now - transitionStart
            if forceAway || gazeIsAway(gazeAngleDeg) {
                state = .disengaged
                return 0.0
            }
            if elapsed >= CorrectionConfig.reEngageDuration {
                state = .engaged
                return 1.0
            }
            return min(1.0, elapsed / CorrectionConfig.reEngageDuration)
        }
    }

    /// Called when tracking is lost, so the next detection starts clean.
    mutating func reset() {
        state = .engaged
        transitionStart = 0
    }
}
